protocol/x/clob/keeper/final_settlement.go

```diff
@@ -0,0 +1,70 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	indexerevents "github.com/dydxprotocol/v4-chain/protocol/indexer/events"
+	"github.com/dydxprotocol/v4-chain/protocol/indexer/indexer_manager"
+	indexershared "github.com/dydxprotocol/v4-chain/protocol/indexer/shared"
+	"github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+)
+
+// MustEnterFinalSettlement holds logic executed when a market transitions to FINAL_SETTLEMENT status.
+// This function will forcefully cancel all stateful open orders for the clob pair.
+func (k Keeper) mustEnterFinalSettlement(ctx sdk.Context, clobPairId types.ClobPairId) {
+	// Forcefully cancel all stateful orders from state for this clob pair
+	k.mustCancelStatefulOrdersForFinalSettlement(ctx, clobPairId)
+
+	// Delete untriggered conditional orders for this clob pair from memory
+	delete(k.UntriggeredConditionalOrders, clobPairId)
+}
+
+// mustCancelStatefulOrdersForFinalSettlement forcefully cancels all stateful orders
+// for the provided ClobPair. These orders will be removed from the memclob in PrepareCheckState.
+func (k Keeper) mustCancelStatefulOrdersForFinalSettlement(ctx sdk.Context, clobPairId types.ClobPairId) {
+	statefulOrders := k.GetAllStatefulOrders(ctx)
+	processProposerMatchesEvents := k.GetProcessProposerMatchesEvents(ctx)
+
+	// This logic is executed in EndBlocker and should not panic. This would be unexpected,
+	// but if it happens we would rather recover and continue if an order fails to be removed from state
+	// rather than halt the chain.
+	safelyRemoveStatefulOrder := func(ctx sdk.Context, orderId types.OrderId) {
+		defer func() {
+			if r := recover(); r != nil {
+				k.Logger(ctx).Error(
+					"mustCancelStatefulOrdersForFinalSettlement: Failed to remove stateful order with OrderId %+v: %v",
+					orderId,
+					r,
+				)
+			}
+		}()
```

Comment:
for my understanding - when would a panic occur?

Reply:
a panic could occur if for some reason the order failed to be found in state (unlikely since we just fetched the whole list of stateful orders), but also from this block. 
```
if len(longTermOrdersExpiringAtTime) != len(updatedStatefulOrdersExpiringAtTime)+1 {
		panic(
			fmt.Sprintf(
				"MustRemoveStatefulOrder: order ID %v is not in state for time %v",
				orderId,
				goodTilBlockTime,
			),
		)
	}
```

I think these are unlikely scenarios, but I had a brief convo with Lucas about this and think it makes sense to recover from panics here so that EndBlocker doesn't panic if a stateful order simply fails to be removed from a final settlement market. If a bug is introduced into that code path we don't want EndBlocker panicking

Reply:
I see. 

In that case, we shouldn't be calling `MustRemoveStatefulOrder` and recover from panics. We should define a `MaybeRemoveStatefulOrder` and not panic at all. 

Slight preference for a new function for cleaner code, but will let you make the call

Reply:
i think I will leave this as is, defining the MaybeRemoveStatefulOrder function would basically involve copy and pasting the entire MustRemoveStatefulOrder function. Seems fine to just recover from the panic honestly

Reply:
Sure, functional-wise it's fine to panic and recover. but recovering from panics that may potentially occur is a code smell and tech debt imo, and may mask some real panics. 

not sure about Go, but IIRC in languages like Java, it's also quite expensive to catch/handle exceptions, so there might be performance implications. 

No need to block on this, let's go ahead and merge this PR. I will probably put out a PR to clean this up.